server/aesdsocket.c

```diff
@@ -77,52 +108,236 @@ void cleanUp(void)
         sockfd = -1;
     }
 
-    if (openConSock != -1) 
+    addr_size = 0;
+
+    struct node * e ;
+    while(!SLIST_EMPTY(&head))
     {
-        shutdown(openConSock,SHUT_RDWR);
-        close(openConSock);
-        openConSock = -1;
+        e = SLIST_FIRST(&head);
+        SLIST_REMOVE(&head,e,node,nodes);
+        pthread_join(e->id, NULL);
+        shutdown(e->socketFd,SHUT_RDWR);
+        close(e->socketFd);
+        e->socketFd = -1;
+        free(e);
+        
     }
 
-    syslog(LOG_INFO,"Caught signal, exiting");
+    pthread_mutex_destroy(&mutex);
+    pthread_mutex_destroy(&trasactionMutex);
+    
+
+    free(thisNode);
+    
+
+    syslog(LOG_INFO,"Caught signal or failed, exiting");
 
     closelog();
     remove(DATA_PATH);
+    exit(exitVal);
 }
 
 void exitSigHandler(int signo) 
 {
     if (signo == SIGINT || signo == SIGTERM) 
     {
-        cleanUp();
-        exit(0);
+        // Setup for main to cleanup!
+        cleanUpTime = true;
+        cleanUp(EXIT_SUCCESS); // not totally safe but passes for now... The problem is cleaningup time stamper quickly for the test
+    }
+}
+
+void *timeStamper () 
+{
+    // chat gpt "This program appends a timestamp in the form “timestamp:time” to the /var/tmp/aesdsocketdata 
+    // file every 10 seconds using the RFC 2822 compliant strftime format. The string includes the year, month, 
+    // day, hour (in 24 hour format), minute, and second representing the system wall clock time."
+    while(!cleanUpTime)
+    {
+        char timestamp[128] ;
+        sleep(DELAY_TO_STAMP);
+
+        pthread_mutex_lock(&trasactionMutex);
+        time_t t = time(NULL);
+        struct tm *tm = localtime(&t);
+        int timeStampfp;
+        strftime(timestamp, sizeof(timestamp), "timestamp: %Y-%m-%d %H:%M:%S\n", tm);
+        timeStampfp = open(DATA_PATH,   O_RDWR | O_CREAT |  O_APPEND, 0664);
+        if(write(timeStampfp,timestamp,sizeof(timestamp)) == -1)
+        {
+            syslog(LOG_INFO,"Failed to write timeStamp");
+            cleanUp(EXIT_FAILURE);
+        }     
+        close(timeStampfp);
+        tm = NULL;
+        pthread_mutex_unlock(&trasactionMutex);
+
+    }
+    cleanUp(EXIT_SUCCESS);
+    return NULL;
+}
+
+
+
+
+
+
+//https://blog.taborkelly.net/programming/c/2016/01/09/sys-queue-example.html
+void *add_to_queue_and_send(void *arg) {
+    struct node* addedNode = (struct node *)arg;
+    pthread_mutex_lock(&mutex);
```

Comment:
@Nicholas-Buckley77 What is mutex locking here? It would be better to add a more descriptive name.

Also, you need to add locking around the rest of the accesses to head as well to make this thread safe, or you need to move this access out of the new thread context so all accesses to head occur from the same thread.